### PR TITLE
fix #42121 fix #78536 fix #103921: Slurs lost on exchange voices and add test for exchange voices

### PR DESCRIPTION
**Pull request by Rockettwo**
I hope I got all Issues on this.

### CODE REVIEW

**Comment by MarcSabatella** [comment]
Looks about right to me.  Just a couple of observations:

- Is it really only slurs that would be affected?  Could be where track2 is meaningful, I guess.  We've changed how this works a few times; I can't remember if any other lines really use track2 or not.

- Instead of calculating s->idx * VOICES over and over, might as well do it once at the top of the loop through staves.

- Nitpick: space before and after the "*" in the above

**Comment by Rockettwo** [comment]
I've updated it as you've suggested in point two and three.

I was thinking about your first point, too and found out that all the lines are affected except the voltas. 
But I'm not quite sure if it would make sense or is to change the track for the other lines, too.
For example for crescendo and decrescendo the track doesn't matter, it's just the start/end tick and the staff.

If I understand it correctly the track2 is the track for the Spanner object and the normal track for the Element object, but correct me if I'm wrong.

So tell me what to do. I would leave it as it is now, because it is not 'better' if we change the track of every line. By the way it would be fairly easy to do so;)

**Comment by MarcSabatella** [comment]
I would say if we can't reproduce a *problem* using other lines (and I couldn't in a few limited attempts) then we shouldn't "fix" them.  But maybe it is worth trying harder to see if we can cause a problem with other lines.

**Comment by mgavioli** [comment]
"_If I understand it correctly the track2 is the track for the Spanner object and the normal track for the Element object, but correct me if I'm wrong._"

As I understand it, `track` is the track of the spanner **starting point** and `track2` is the track of the spanner **end point**, to allow spanners which start in a track and end in another.

**Comment by lasconic** [comment]
@mgavioli is right.
Regarding this PR, it would be good to have some tests with it and more globally for exchange voices.
A new directory mtest/libmscore/exchangevoices would be good.

File: libmscore/undo.cpp [hunk 1]
```diff
@@ -715,6 +718,20 @@ void Score::undoExchangeVoice(Measure* measure, int v1, int v2, int staff1, int
       for (Staff* s : sl) {
             Measure* m = s->score()->tick2measure(tick);
             undo(new ExchangeVoice(m, v1, v2, s->idx()));
+            for (auto i = spanners.begin(); i < spanners.end(); i++) {
+                  Spanner* sp = i->value;
+                  int off = s->idx() * VOICES;
```
Comment (Jojo-Schmitz) [inline libmscore/undo.cpp]: I think you could move that up into the outer loop, would certainly be cheaper

**Comment by Rockettwo** [comment]
Ok, makes sense.
So I haven't ever written a test, but I'll try to do so.
Or should I make this in a separate PR?

**Comment by Jojo-Schmitz** [comment]
Not a separate PR, at most a separate commit to this one

**Comment by MarcSabatella** [comment]
Sorry I wasn't more clear.

Indeed, track for a spanner is the track of the start element, track2 is the track of the end element for spanners that can cross voices.  The issue for me is, I can't remember which spanners are allowed to cross voices.  Clearly, slurs can atach to notes in different voices - we go to some trouble to allow that.  But most spanners don't attach to notes, so there isn't a reason to set track2 to anything in particular.

In theory, I could see track2 being used in a hairpin or other spanner to allow it to end with a note in a different voice from the voice it started in, but if there is a way to create that, I don't know what it is.

**Comment by Rockettwo** [comment]
> Clearly, slurs can attach to notes in different voices

Regarding this I'll have to change the commit again, because I have to check track and track2 separately.

@lasconic I've added the directory but I'm not sure what I should test. I did attach a test for slurs in case they are in the same voice and in case they have different start and end voices.
What else do you suggest?

I'm not quite sure if the test is really working because on my computer it always says failure location (Win 10)... Is this a known issue?

**Comment by mgavioli** [comment]
@MarcSabatella : at least glissandi too are allowed to cross voices, even to cross staves.

**Comment by Jojo-Schmitz** [comment]
Ththe test Fails on Travis too. I believe you'd deed to create the test files file in -t mode, and then what the diff Output of that test on Travis

**Comment by Jojo-Schmitz** [comment]
The test fails on Travis too. I believe you'd deed to create the test files file in -t mode, and then watch the diff output of that test on Travis
See https://travis-ci.org/musescore/MuseScore/jobs/120352394#L4605

**Comment by Jojo-Schmitz** [comment]
OK, very few Problems left:
```
16/48 Test #16: tst_exchangevoices ...............***Failed    0.73 sec
--- exchangevoices-slurs.mscx	2016-04-04 05:58:06.746719672 +0000
+++ /home/travis/build/musescore/MuseScore/mtest/libmscore/exchangevoices/exchangevoices-slurs-ref.mscx	2016-04-04 05:45:03.737509493 +0000
@@ -195,7 +195,6 @@
           <track>1</track>
           <durationType>eighth</durationType>
           <Beam>2</Beam>
-          <Slur type="stop" id="2"/>
           <Note>
             <track>1</track>
             <pitch>72</pitch>
@@ -273,4 +272,4 @@
         </Measure>
       </Staff>
     </Score>
-  </museScore>
+  </museScore>
\ No newline at end of file
   <diff -u exchangevoices-slurs.mscx /home/travis/build/musescore/MuseScore/mtest/libmscore/exchangevoices/exchangevoices-slurs-ref.mscx failed
````

**Comment by Rockettwo** [comment]
Okay, I think this is an real issue, I'll take a look on this later.
The actual commit should do it all correctly, but it's a bit cheated, so don't merge this already...

File: libmscore/undo.h [hunk 2]
```diff
@@ -479,6 +479,7 @@ class ExchangeVoice : public UndoCommand {
       int val1, val2;
       int staff;
 
+
```
Comment (Jojo-Schmitz) [inline libmscore/undo.h]: as per https://musescore.org/en/developers-handbook/musescore-coding-rules#Indentation-%26-Whitespaces:
•Always use only one blank line.

**Comment by Rockettwo** [comment]
I switched my development OS to Linux and got everything to work. 
I've had some problems with slurs which last over multiple measures and the selection is set in between. But I couldn't figure out how to test this case in the test. 
The last problem which I actually don't know how to fix is the undo progress, because an undo command does actually not really undo the last step but is only switching the voices back, this is what it is expected to but the beam settings aren't handled correctly. 
This is the actual test case:
scsh fir

This is the output, if the selected voices are exchanged:
scshdo

This is the result of the undo command
scsh undo

So the voices are switched back but the beam settings are not restored.
How can we fix this?
This causes also a failure of the exchangevoices test, but to give you the opportunity to reproduce this the actual reference is a copy of a previous test output.

File: libmscore/undo.h (hunk 2 above)
Comment (Rockettwo) [inline libmscore/undo.h]: okay is just done

**Comment by Rockettwo** [comment]
If I compare the XML data of the scores it seems that there are more beam settings saved when I run the test as when I simply save it with Musescore. So it seems the actual compare scores are already correct.
The undo issue may be fixed when we'll also switch the beam settings!?

Please give me some feedback.

**Comment by lasconic** [comment]
beam settings are only saved in the score if you run MuseScore with -t.
See https://github.com/musescore/MuseScore/tree/master/mtest#test-case-conventions about creating test cases.

**Comment by Jojo-Schmitz** [comment]
Rebase needed

**Comment by Rockettwo** [comment]
Rebase done.

But I still don't know how to fix the undo issue...

**Comment by Rockettwo** [comment]
Okay, seems it is failing on travis due to the beam settings, but not on my PC.
Does Travis have a special command line option to call the test itself?

File: mtest/libmscore/exchangevoices/exchangevoices-slurs-ref.mscx [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    <metaTag name="composer"></metaTag>
+    <metaTag name="copyright"></metaTag>
+    <metaTag name="lyricist"></metaTag>
+    <metaTag name="movementNumber"></metaTag>
+    <metaTag name="movementTitle"></metaTag>
+    <metaTag name="poet"></metaTag>
+    <metaTag name="source"></metaTag>
+    <metaTag name="translator"></metaTag>
+    <metaTag name="workNumber"></metaTag>
+    <metaTag name="workTitle">HairPin-Voice test</metaTag>
```
Comment (Jojo-Schmitz) [inline mtest/libmscore/exchangevoices/exchangevoices-slurs-ref.mscx]: Not Slur test? Maybe better no workTitle at all?

**Comment by Jojo-Schmitz** [comment]
ISTR that yesterday the Travis test showed sussess, did you cange something in the code since then? What exactly? Did you redo the test files after that change?

**Comment by Rockettwo** [comment]
Oh I  got my mistake, I always used an old version of the tst_exchangevoices, because I rebuild the mtest, but just run the tst_exchangevoices. Sorry for this! Correct version is coming.

**Comment by Jojo-Schmitz** [comment]
Still an mtest failure

**Comment by Jojo-Schmitz** [comment]
Ah, finally Travis is happy with it :-)

**Comment by Rockettwo** [comment]
Yeah, I think it was because I used the newest version of Qt with gcc 5.5.1 but I'm not sure...

**Comment by lasconic** [comment]
This PR is still mergeable, but the tests were failing. I rebased it and fixed the tests. Also it seems to work equally well with the current measure instead of passing it as a parameter https://github.com/musescore/MuseScore/pull/2517/files#diff-2e7c83f7c549a471a12a17756ff33514R2616

**Pull request closed without merging**